neuralmonkey/decoders/transformer.py

```diff
@@ -198,15 +225,25 @@ def layer(self, level: int, inputs: tf.Tensor,
     def train_logits(self) -> tf.Tensor:
         last_layer = self.layer(self.depth, self.embedded_train_inputs,
                                 tf.transpose(self.train_mask))
-        # matmul with output matrix
         # t_states shape: (batch, time, channels)
         # dec_w shape: (channels, vocab)
-        logits_unbiased = tf.nn.conv1d(
-            last_layer.temporal_states, tf.expand_dims(self.decoding_w, 0), 1,
-            "SAME")
+        last_layer_shape = tf.shape(last_layer.temporal_states)
+        last_layer_states = tf.reshape(
+            last_layer.temporal_states,
+            [-1, last_layer_shape[-1]])
+
+        # TODO: Add bias after matmul by embedding_matrix?
 
+        # Reusing input embedding matrix for generating logits
+        # significantly reduces the overall size of the model.
+        # See: https://arxiv.org/pdf/1608.05859.pdf
+        #
         # shape (batch, time, vocab)
-        logits = logits_unbiased + tf.reshape(self.decoding_b, [1, 1, -1])
+        logits = tf.reshape(
+            tf.matmul(
+                last_layer_states,
+                tf.transpose(self.embedding_matrix)),
+            [last_layer_shape[0], last_layer_shape[1], len(self.vocabulary)])
```

Comment:
Když se sdílí ta matice, tak tam není bias?

Reply:
Ja si myslim, ze by tam nemel byt (jsou to prece jen extra promenne). Podivam se do toho clanku, jestli to tam neni popsane.

Reply:
Bias nad slovníkem neuškodí, ne?

Reply:
Pokud nic neprida, tak to jsou zbytecne promenne navic (viz vyse).